File: src/main/java/com/snowflake/snowpark_java/Updatable.java
package com.snowflake.snowpark_java;

import com.snowflake.snowpark.internal.JavaUtils;
import java.util.HashMap;
import java.util.Map;

/**
 * Represents a lazily-evaluated Updatable. It extends {@code DataFrame} so all {@code DataFrame}
 * operations can be applied on it.
 *
 * <p>'''Creating an Updatable'''
 *
 * <p>You can create an Updatable by calling {@code Session.table} with the name of the Updatable.
 *
 * @since 1.1.0
 */
public class Updatable extends DataFrame {
  private final com.snowflake.snowpark.Updatable updatable;

  Updatable(com.snowflake.snowpark.Updatable updatable) {
    super(updatable);
    this.updatable = updatable;
  }

  /**
   * Updates all rows in the Updatable with specified assignments and returns a {@code
   * UpdateResult}, representing number of rows modified and number of multi-joined rows modified.
   *
   * <p>For example:
   *
   * <pre>{@code
   * Map<Column, Column> assignments = new HashMap<>;
   * assignments.put(Functions.col("col1"), Functions.lit(1);
   * // Assign value 1 to column col1 in all rows in updatable.
   * session.table("tableName").update(assignment);
   * }</pre>
   *
   * @since 1.1.0
   * @param assignments A map contains the column being updated and the new value.
   * @return {@code UpdateResult}
   */
  public UpdateResult update(Map<Column, Column> assignments) {
    return new UpdateResult(JavaUtils.updatable_update(toScalaColumnMap(assignments), updatable));
  }

  /**
   * Updates all rows in the updatable with specified assignments and returns a {@code
   * UpdateResult}, representing number of rows modified and number of multi-joined rows modified.
   *
   * <p>For example:
   *
   * <pre>{@code
   * Map<String, Column> assignments = new HashMap<>;
   * assignments.put("col1", Functions.lit(1);
   * // Assign value 1 to column col1 in all rows in updatable.
   * session.table("tableName").updateColumn(assignment);
   * }</pre>
   *
   * @since 1.1.0
   * @param assignments A map contains the column being updated and the new value.
   * @return {@code UpdateResult}
   */
  public UpdateResult updateColumn(Map<String, Column> assignments) {
    return new UpdateResult(
        JavaUtils.updatable_updateColumn(toScalaStringColumnMap(assignments), updatable));
  }

  /**
   * Updates all rows in the updatable that satisfy specified condition with specified assignments
   * and returns a {@code UpdateResult}, representing number of rows modified and number of
   * multi-joined rows modified.
   *
   * <p>For example:
   *
   * <pre>{@code
   * Map<Column, Column> assignments = new HashMap<>;
   * assignments.put(Functions.col("col1"), Functions.lit(1);
   * // Assign value 1 to column col1 in the rows if col2 is true in updatable.
   * session.table("tableName").updateColumn(assignment,
   *   Functions.col("col2").equal_to(Functions.lit(true)));
   * }</pre>
   *
   * @since 1.1.0
   * @param assignments A map contains the column being updated and the new value.
   * @param condition The condition of the Column being updated.
   * @return {@code UpdateResult}
   */
  public UpdateResult update(Map<Column, Column> assignments, Column condition) {
    return new UpdateResult(
        JavaUtils.updatable_update(
            toScalaColumnMap(assignments), condition.toScalaColumn(), updatable));
  }

  /**
   * Updates all rows in the updatable that satisfy specified condition with specified assignments
   * and returns a {@code UpdateResult}, representing number of rows modified and number of
   * multi-joined rows modified.
   *
   * <p>For example:
   *
   * <pre>{@code
   * Map<String, Column> assignments = new HashMap<>;
   * assignments.put("col1", Functions.lit(1);
   * // Assign value 1 to column col1 in the rows if col3 is true in updatable.
   * session.table("tableName").updateColumn(assignment,
   *   Functions.col("col2").equal_to(Functions.lit(true)));
   * }</pre>
   *
   * @since 1.1.0
   * @param assignments A map contains the column being updated and the new value.
   * @param condition The condition of the Column being updated.
   * @return {@code UpdateResult}
   */
  public UpdateResult updateColumn(Map<String, Column> assignments, Column condition) {
    return new UpdateResult(
        JavaUtils.updatable_updateColumn(
            toScalaStringColumnMap(assignments), condition.toScalaColumn(), updatable));
  }

  /**
   * Updates all rows in the updatable that satisfy specified condition where condition includes
   * columns in other {@code DataFrame}, and returns a {@code UpdateResult}, representing number of
   * rows modified and number of multi-joined rows modified.
   *
   * <p>For example:
   *
   * <pre>{@code
   * Map<Column, Column> assignments = new HashMap<>;
   * assignments.put(Functions.col("col1"), Functions.lit(1);
   * session.table("tableName").update(assignment,
   *   Functions.col("col2").equal_to(df.col("col_a")), df);
   * }</pre>
   *
   * @since 1.1.0
   * @param assignments A map contains the column being updated and the new value.
   * @param condition The condition of the Column being updated.
   * @param sourceData Another DataFrame being joined.
   * @return {@code UpdateResult}
   */
  public UpdateResult update(
      Map<Column, Column> assignments, Column condition, DataFrame sourceData) {
    return new UpdateResult(
        JavaUtils.updatable_update(
            toScalaColumnMap(assignments),
            condition.toScalaColumn(),
            sourceData.getScalaDataFrame(),
            updatable));
  }

  /**
   * Updates all rows in the updatable that satisfy specified condition where condition includes
   * columns in other {@code DataFrame}, and returns a {@code UpdateResult}, representing number of
   * rows modified and number of multi-joined rows modified.
   *
   * <p>For example:
   *
   * <pre>{@code
   * Map<String, Column> assignments = new HashMap<>;
   * assignments.put("col1", Functions.lit(1);
   * session.table("tableName").updateColumn(assignment,
   *   Functions.col("col2").equal_to(df.col("col_a")), df);
   * }</pre>
   *
   * @since 1.1.0
   * @param assignments A map contains the column being updated and the new value.
   * @param condition The condition of the Column being updated.
   * @param sourceData Another DataFrame being joined.
   * @return {@code UpdateResult}
   */
  public UpdateResult updateColumn(
      Map<String, Column> assignments, Column condition, DataFrame sourceData) {
    return new UpdateResult(
        JavaUtils.updatable_updateColumn(
            toScalaStringColumnMap(assignments),
            condition.toScalaColumn(),
            sourceData.getScalaDataFrame(),
            updatable));
  }

  /**
   * Deletes all rows in the updatable and returns a {@code DeleteResult}, representing number of
   * rows deleted.
   *
   * @since 1.1.0
   * @return {@code DeleteResult}
   */
  public DeleteResult delete() {
    return new DeleteResult(updatable.delete());
  }

  /**
   * Deletes all rows in the updatable that satisfy specified condition and returns a {@code
   * DeleteResult}, representing number of rows deleted.
   *
   * <p>For example:
   *
   * <pre>{@code
   * session.table("tableName").delete(Functions.col("col1").equal_to(Functions.lit(1)));
   * }</pre>
   *
   * @since 1.1.0
   * @param condition The condition expression
   * @return {@code DeleteResult}
   */
  public DeleteResult delete(Column condition) {
    return new DeleteResult(updatable.delete(condition.toScalaColumn()));
  }

  /**
   * Deletes all rows in the updatable that satisfy specified condition where condition includes
   * columns in other {@code DataFrame}, and returns a {@code DeleteResult}, representing number of
   * rows deleted.
   *
   * <p>For example:
   *
   * <pre>{@code
   * session.table(tableName).delete(Functions.col("col1").equal_to(df.col("col2")), df);
   * }</pre>
   *
   * @since 1.1.0
   * @param condition The condition expression
   * @param sourceData The source DataFrame
   * @return {@code DeleteResult}
   */
  public DeleteResult delete(Column condition, DataFrame sourceData) {
    return new DeleteResult(
        updatable.delete(condition.toScalaColumn(), sourceData.getScalaDataFrame()));
  }

  /**
   * Initiates a merge action for this updatable with {@code DataFrame} source on specified join
   * expression. Returns a {@code MergeBuilder} which provides APIs to define merge clauses.
   *
   * @since 1.1.0
   * @param source The source DataFrame
   * @param joinExpr The join expression
   * @return {@code MergeBuilder}
   */
  public MergeBuilder merge(DataFrame source, Column joinExpr) {
    return new MergeBuilder(
        this.updatable.merge(source.getScalaDataFrame(), joinExpr.toScalaColumn()));
  }

  /**
   * Returns a clone of this Updatable.
   *
   * @return {@code Updatable}
   * @since 1.1.0
   */
  @Override
  public Updatable clone() {
    // invoke super.clone to remove compiler warning
    super.clone();
    return new Updatable(updatable.clone());
  }

  /**
   * Returns an UpdatableAsyncActor object that can be used to execute Updatable actions
   * asynchronously.
   *
   * @since 1.2.0
   * @return A UpdatableAsyncActor object
   */
  @Override
  public UpdatableAsyncActor async() {
    return new UpdatableAsyncActor(this);
  }

  com.snowflake.snowpark.Updatable getScalaUpdatable() {
    return this.updatable;
  }

  static Map<com.snowflake.snowpark.Column, com.snowflake.snowpark.Column> toScalaColumnMap(
      Map<Column, Column> input) {
    Map<com.snowflake.snowpark.Column, com.snowflake.snowpark.Column> result = new HashMap<>();
    for (Map.Entry<Column, Column> entry : input.entrySet()) {
      result.put(entry.getKey().toScalaColumn(), entry.getValue().toScalaColumn());
    }
    return result;
  }

  static Map<String, com.snowflake.snowpark.Column> toScalaStringColumnMap(
      Map<String, Column> input) {
    Map<String, com.snowflake.snowpark.Column> result = new HashMap<>();
    for (Map.Entry<String, Column> entry : input.entrySet()) {
      result.put(entry.getKey(), entry.getValue().toScalaColumn());
    }
    return result;
  }
}

File: src/main/scala/com/snowflake/snowpark/Updatable.scala
package com.snowflake.snowpark

import com.snowflake.snowpark.internal.{Logging, OpenTelemetry}
import com.snowflake.snowpark.internal.analyzer._

import scala.reflect.ClassTag

private[snowpark] object Updatable extends Logging {
  def apply(tableName: String, session: Session): Updatable =
    new Updatable(tableName, session, DataFrame.methodChainCache.value)

  private[snowpark] def getUpdateResult(rows: Array[Row]): UpdateResult =
    UpdateResult(rows.head.getLong(0), rows.head.getLong(1))

  private[snowpark] def getDeleteResult(rows: Array[Row]): DeleteResult =
    DeleteResult(rows.head.getLong(0))

}

/**
 * Result of updating rows in an Updatable
 *
 * @since 0.7.0
 */
case class UpdateResult(rowsUpdated: Long, multiJoinedRowsUpdated: Long)

/**
 * Result of deleting rows in an Updatable
 *
 * @since 0.7.0
 */
case class DeleteResult(rowsDeleted: Long)

/**
 * Represents a lazily-evaluated Updatable. It extends [[DataFrame]] so all
 * [[DataFrame]] operations can be applied on it.
 *
 * '''Creating an Updatable'''
 *
 * You can create an Updatable by calling [[Session.table(name* session.table]] with the name of
 * the Updatable.
 *
 * Example 1: Creating a Updatable by reading a table.
 * {{{
 *   val dfPrices = session.table("itemsdb.publicschema.prices")
 * }}}
 *
 * @groupname actions Actions
 * @groupname basic Basic DataFrame Functions
 *
 * @since 0.7.0
 */
class Updatable private[snowpark] (
    private[snowpark] val tableName: String,
    override private[snowpark] val session: Session,
    override private[snowpark] val methodChain: Seq[String])
    extends DataFrame(
      session,
      session.analyzer.resolve(UnresolvedRelation(tableName)),
      methodChain) {

  /**
   * Updates all rows in the Updatable with specified assignments and returns a [[UpdateResult]],
   * representing number of rows modified and number of multi-joined rows modified.
   *
   * For example:
   * {{{
   *   updatable.update(Map(col("b") -> lit(0)))
   * }}}
   *
   * Assign value 0 to column b in all rows in updatable.
   *
   * {{{
   *   updatable.update(Map(col("c") -> (col("a") + col("b"))))
   * }}}
   *
   * Assign the sum of column a and column b to column c in all rows in updatable
   *
   * @group actions
   * @since 0.7.0
   * @return [[UpdateResult]]
   */
  def update(assignments: Map[Column, Column]): UpdateResult = action("update", 2) {
    val newDf = getUpdateDataFrameWithColumn(assignments, None, None)
    Updatable.getUpdateResult(newDf.collect())
  }

  /**
   * Updates all rows in the updatable with specified assignments and returns a [[UpdateResult]],
   * representing number of rows modified and number of multi-joined rows modified.
   *
   * For example:
   * {{{
   *   updatable.update(Map("b" -> lit(0)))
   * }}}
   *
   * Assign value 0 to column b in all rows in updatable.
   *
   * {{{
   *   updatable.update(Map("c" -> (col("a") + col("b"))))
   * }}}
   *
   * Assign the sum of column a and column b to column c in all rows in updatable
   *
   * @group actions
   * @since 0.7.0
   * @return [[UpdateResult]]
   */
  def update[T: ClassTag](assignments: Map[String, Column]): UpdateResult = action("update", 2) {
    val newDf = getUpdateDataFrameWithString(assignments, None, None)
    Updatable.getUpdateResult(newDf.collect())
  }

  /**
   * Updates all rows in the updatable that satisfy specified condition with specified assignments
   * and returns a [[UpdateResult]], representing number of rows modified and number of
   * multi-joined rows modified.
   *
   * For example:
   * {{{
   *   updatable.update(Map(col("b") -> lit(0)), col("a") === 1)
   * }}}
   *
   * Assign value 0 to column b in all rows where column a has value 1.
   *
   * @group actions
   * @since 0.7.0
   * @return [[UpdateResult]]
   */
  def update(assignments: Map[Column, Column], condition: Column): UpdateResult =
    action("update", 2) {
      val newDf = getUpdateDataFrameWithColumn(assignments, Some(condition), None)
      Updatable.getUpdateResult(newDf.collect())
    }

  /**
   * Updates all rows in the updatable that satisfy specified condition with specified assignments
   * and returns a [[UpdateResult]], representing number of rows modified and number of
   * multi-joined rows modified.
   *
   * For example:
   * {{{
   *   updatable.update(Map("b" -> lit(0)), col("a") === 1)
   * }}}
   *
   * Assign value 0 to column b in all rows where column a has value 1.
   *
   * @group actions
   * @since 0.7.0
   * @return [[UpdateResult]]
   */
  def update[T: ClassTag](assignments: Map[String, Column], condition: Column): UpdateResult =
    action("update", 2) {
      val newDf = getUpdateDataFrameWithString(assignments, Some(condition), None)
      Updatable.getUpdateResult(newDf.collect())
    }

  /**
   * Updates all rows in the updatable that satisfy specified condition where condition includes
   * columns in other [[DataFrame]], and returns a [[UpdateResult]], representing number of rows
   * modified and number of multi-joined rows modified.
   *
   * For example:
   * {{{
   *   t1.update(Map(col("b") -> lit(0)), t1("a") === t2("a"), t2)
   * }}}
   *
   * Assign value 0 to column b in all rows in t1 where column a in t1 equals column a in t2.
   *
   * @group actions
   * @since 0.7.0
   * @return [[UpdateResult]]
   */
  def update(
      assignments: Map[Column, Column],
      condition: Column,
      sourceData: DataFrame): UpdateResult = action("update", 2) {
    val newDf = getUpdateDataFrameWithColumn(assignments, Some(condition), Some(sourceData))
    Updatable.getUpdateResult(newDf.collect())
  }

  /**
   * Updates all rows in the updatable that satisfy specified condition where condition includes
   * columns in other [[DataFrame]], and returns a [[UpdateResult]], representing number of rows
   * modified and number of multi-joined rows modified.
   *
   * For example:
   * {{{
   *   t1.update(Map("b" -> lit(0)), t1("a") === t2("a"), t2)
   * }}}
   *
   * Assign value 0 to column b in all rows in t1 where column a in t1 equals column a in t2.
   *
   * @group actions
   * @since 0.7.0
   * @return [[UpdateResult]]
   */
  def update[T: ClassTag](
      assignments: Map[String, Column],
      condition: Column,
      sourceData: DataFrame): UpdateResult = action("update", 2) {
    val newDf = getUpdateDataFrameWithString(assignments, Some(condition), Some(sourceData))
    Updatable.getUpdateResult(newDf.collect())
  }

  private[snowpark] def getUpdateDataFrameWithString(
      assignments: Map[String, Column],
      condition: Option[Column],
      sourceData: Option[DataFrame]): DataFrame =
    getUpdateDataFrameWithColumn(
      assignments.map { case (k, v) => (col(k), v) },
      condition,
      sourceData)

  private[snowpark] def getUpdateDataFrameWithColumn(
      assignments: Map[Column, Column],
      condition: Option[Column],
      sourceData: Option[DataFrame]): DataFrame = {
    session.conn.telemetry.reportActionUpdate()
    withPlan(
      TableUpdate(
        tableName,
        assignments.map { case (k, v) => (k.expr, v.expr) },
        condition.map(_.expr),
        sourceData.map(disambiguate(this, _, JoinType("left"), Seq.empty)._2.plan)))
  }

  /**
   * Deletes all rows in the updatable and returns a [[DeleteResult]], representing number of rows
   * deleted.
   *
   * For example:
   * {{{
   *   updatable.delete()
   * }}}
   *
   * Deletes all rows in updatable.
   *
   * @group actions
   * @since 0.7.0
   * @return [[DeleteResult]]
   */
  def delete(): DeleteResult = action("delete") {
    val newDf = getDeleteDataFrame(None, None)
    Updatable.getDeleteResult(newDf.collect())
  }

  /**
   * Deletes all rows in the updatable that satisfy specified condition and returns a
   * [[DeleteResult]], representing number of rows deleted.
   *
   * For example:
   * {{{
   *   updatable.delete(col("a") === 1)
   * }}}
   *
   * Deletes all rows where column a has value 1.
   *
   * @group actions
   * @since 0.7.0
   * @return [[DeleteResult]]
   */
  def delete(condition: Column): DeleteResult = action("delete") {
    val newDf = getDeleteDataFrame(Some(condition), None)
    Updatable.getDeleteResult(newDf.collect())
  }

  /**
   * Deletes all rows in the updatable that satisfy specified condition where condition includes
   * columns in other [[DataFrame]], and returns a [[DeleteResult]], representing number of rows
   * deleted.
   *
   * For example:
   * {{{
   *   t1.delete(t1("a") === t2("a"), t2)
   * }}}
   *
   * Deletes all rows in t1 where column a in t1 equals column a in t2.
   *
   * @group actions
   * @since 0.7.0
   * @return [[DeleteResult]]
   */
  def delete(condition: Column, sourceData: DataFrame): DeleteResult = action("delete") {
    val newDf = getDeleteDataFrame(Some(condition), Some(sourceData))
    Updatable.getDeleteResult(newDf.collect())
  }

  private[snowpark] def getDeleteDataFrame(
      condition: Option[Column],
      sourceData: Option[DataFrame]): DataFrame = {
    session.conn.telemetry.reportActionDelete()
    withPlan(
      TableDelete(
        tableName,
        condition.map(_.expr),
        sourceData.map(disambiguate(this, _, JoinType("left"), Seq.empty)._2.plan)))
  }

  /**
   * Initiates a merge action for this updatable with [[DataFrame]] source on specified
   * join expression. Returns a [[MergeBuilder]] which provides APIs to define merge clauses.
   *
   * For example:
   * {{{
   *   target.merge(source, target("id") === source("id"))
   * }}}
   *
   * Initiates a merge action for target with source where the expression target.id = source.id
   * is used to join target and source.
   *
   * @group actions
   * @since 0.7.0
   * @return [[MergeBuilder]]
   */
  def merge(source: DataFrame, joinExpr: Column): MergeBuilder = {
    session.conn.telemetry.reportActionMerge()
    MergeBuilder(
      this,
      disambiguate(this, source, JoinType("left"), Seq.empty)._2,
      joinExpr,
      Seq.empty,
      inserted = false,
      updated = false,
      deleted = false)
  }

  /**
   * Returns a clone of this Updatable.
   *
   * @return A [[Updatable]]
   * @since 0.10.0
   * @group basic
   */
  override def clone: Updatable = action("clone", 2) {
    new Updatable(tableName, session, Seq())
  }

  /**
   * Returns an [[UpdatableAsyncActor]] object that can be used to execute
   * Updatable actions asynchronously.
   *
   * Example:
   * {{{
   *   val updatable = session.table(tableName)
   *   val asyncJob = updatable.async.update(Map(col("b") -> lit(0)), col("a") === 1)
   *   // At this point, the thread is not blocked. You can perform additional work before
   *   // calling asyncJob.getResult() to retrieve the results of the action.
   *   // NOTE: getResult() is a blocking call.
   *   val updateResult = asyncJob.getResult()
   * }}}
   *
   * @since 0.11.0
   * @return A [[UpdatableAsyncActor]] object
   */
  override def async: UpdatableAsyncActor = new UpdatableAsyncActor(this)

  @inline override protected def action[T](funcName: String)(func: => T): T = {
    val isScala: Boolean = this.session.conn.isScalaAPI
    OpenTelemetry.action("Updatable", funcName, methodChainString, isScala)(func)
  }

  @inline protected def action[T](funcName: String, javaOffset: Int)(func: => T): T = {
    val isScala: Boolean = this.session.conn.isScalaAPI
    OpenTelemetry.action("Updatable", funcName, methodChainString, isScala, javaOffset)(func)
  }

}

/**
 * Provides APIs to execute Updatable actions asynchronously.
 *
 * @since 0.11.0
 */
class UpdatableAsyncActor private[snowpark] (updatable: Updatable)
    extends DataFrameAsyncActor(updatable) {

  /**
   * Executes `Updatable.update` asynchronously.
   *
   * @return A [[TypedAsyncJob]] object that you can use to check the status of the action
   *         and get the results.
   * @since 0.11.0
   */
  def update(assignments: Map[Column, Column]): TypedAsyncJob[UpdateResult] =
    action("update", 2) {
      val newDf = updatable.getUpdateDataFrameWithColumn(assignments, None, None)
      updatable.session.conn.executeAsync[UpdateResult](newDf.snowflakePlan)
    }

  /**
   * Executes `Updatable.update` asynchronously.
   *
   * @return A [[TypedAsyncJob]] object that you can use to check the status of the action
   *         and get the results.
   * @since 0.11.0
   */
  def update[T: ClassTag](assignments: Map[String, Column]): TypedAsyncJob[UpdateResult] =
    action("update", 2) {
      val newDf = updatable.getUpdateDataFrameWithString(assignments, None, None)
      updatable.session.conn.executeAsync[UpdateResult](newDf.snowflakePlan)
    }

  /**
   * Executes `Updatable.update` asynchronously.
   *
   * @return A [[TypedAsyncJob]] object that you can use to check the status of the action
   *         and get the results.
   * @since 0.11.0
   */
  def update(assignments: Map[Column, Column], condition: Column): TypedAsyncJob[UpdateResult] =
    action("update", 2) {
      val newDf = updatable.getUpdateDataFrameWithColumn(assignments, Some(condition), None)
      updatable.session.conn.executeAsync[UpdateResult](newDf.snowflakePlan)
    }

  /**
   * Executes `Updatable.update` asynchronously.
   *
   * @return A [[TypedAsyncJob]] object that you can use to check the status of the action
   *         and get the results.
   * @since 0.11.0
   */
  def update[T: ClassTag](
      assignments: Map[String, Column],
      condition: Column): TypedAsyncJob[UpdateResult] =
    action("update", 2) {
      val newDf = updatable.getUpdateDataFrameWithString(assignments, Some(condition), None)
      updatable.session.conn.executeAsync[UpdateResult](newDf.snowflakePlan)
    }

  /**
   * Executes `Updatable.update` asynchronously.
   *
   * @return A [[TypedAsyncJob]] object that you can use to check the status of the action
   *         and get the results.
   * @since 0.11.0
   */
  def update(
      assignments: Map[Column, Column],
      condition: Column,
      sourceData: DataFrame): TypedAsyncJob[UpdateResult] = action("update", 2) {
    val newDf =
      updatable.getUpdateDataFrameWithColumn(assignments, Some(condition), Some(sourceData))
    updatable.session.conn.executeAsync[UpdateResult](newDf.snowflakePlan)
  }

  /**
   * Executes `Updatable.update` asynchronously.
   *
   * @return A [[TypedAsyncJob]] object that you can use to check the status of the action
   *         and get the results.
   * @since 0.11.0
   */
  def update[T: ClassTag](
      assignments: Map[String, Column],
      condition: Column,
      sourceData: DataFrame): TypedAsyncJob[UpdateResult] = action("update", 2) {
    val newDf =
      updatable.getUpdateDataFrameWithString(assignments, Some(condition), Some(sourceData))
    updatable.session.conn.executeAsync[UpdateResult](newDf.snowflakePlan)
  }

  /**
   * Executes `Updatable.delete` asynchronously.
   *
   * @return A [[TypedAsyncJob]] object that you can use to check the status of the action
   *         and get the results.
   * @since 0.11.0
   */
  def delete(): TypedAsyncJob[DeleteResult] = action("delete") {
    val newDf = updatable.getDeleteDataFrame(None, None)
    updatable.session.conn.executeAsync[DeleteResult](newDf.snowflakePlan)
  }

  /**
   * Executes `Updatable.delete` asynchronously.
   *
   * @return A [[TypedAsyncJob]] object that you can use to check the status of the action
   *         and get the results.
   * @since 0.11.0
   */
  def delete(condition: Column): TypedAsyncJob[DeleteResult] = action("delete") {
    val newDf = updatable.getDeleteDataFrame(Some(condition), None)
    updatable.session.conn.executeAsync[DeleteResult](newDf.snowflakePlan)
  }

  /**
   * Executes `Updatable.delete` asynchronously.
   *
   * @return A [[TypedAsyncJob]] object that you can use to check the status of the action
   *         and get the results.
   * @since 0.11.0
   */
  def delete(condition: Column, sourceData: DataFrame): TypedAsyncJob[DeleteResult] =
    action("delete") {
      val newDf = updatable.getDeleteDataFrame(Some(condition), Some(sourceData))
      updatable.session.conn.executeAsync[DeleteResult](newDf.snowflakePlan)
    }

  @inline override protected def action[T](funcName: String)(func: => T): T = {
    val isScala: Boolean = updatable.session.conn.isScalaAPI
    OpenTelemetry.action(
      "UpdatableAsyncActor",
      funcName,
      updatable.methodChainString + ".async",
      isScala)(func)
  }

  @inline protected def action[T](funcName: String, javaOffset: Int)(func: => T): T = {
    val isScala: Boolean = updatable.session.conn.isScalaAPI
    OpenTelemetry.action(
      "UpdatableAsyncActor",
      funcName,
      updatable.methodChainString + ".async",
      isScala,
      javaOffset)(func)
  }
}
